File: packages/clients/js/src/diagnostic.ts
import * as tg from "./index.ts";

export type Diagnostic = {
	location: tg.Module.Location | null;
	message: string;
	severity: Diagnostic.Severity;
};

export namespace Diagnostic {
	export type Severity = "error" | "warning" | "info" | "hint";

	export type Data = {
		location?: tg.Module.Location.Data | null;
		message: string;
		severity: Diagnostic.Severity;
	};

	export let toData = (value: tg.Diagnostic): tg.Diagnostic.Data => {
		let data: tg.Diagnostic.Data = {
			message: value.message,
			severity: value.severity,
		};
		if (value.location !== undefined && value.location !== null) {
			data.location = tg.Module.Location.toData(value.location);
		}
		return data;
	};

	export let fromData = (data: tg.Diagnostic.Data): tg.Diagnostic => {
		let diagnostic: tg.Diagnostic = {
			location:
				data.location !== undefined && data.location !== null
					? tg.Module.Location.fromData(data.location)
					: null,
			message: data.message,
			severity: data.severity,
		};
		return diagnostic;
	};

	export let children = (value: tg.Diagnostic): Array<tg.Object> => {
		if (value.location !== undefined && value.location !== null) {
			return tg.Module.Location.children(value.location);
		} else {
			return [];
		}
	};

	export namespace Data {
		export let children = (data: tg.Diagnostic.Data): Array<tg.Object.Id> => {
			if (data.location !== undefined && data.location !== null) {
				return tg.Module.Location.Data.children(data.location);
			} else {
				return [];
			}
		};

		export let withoutTokens = (
			data: tg.Diagnostic.Data,
		): tg.Diagnostic.Data => {
			if (data.location !== undefined && data.location !== null) {
				data.location = tg.Module.Location.Data.withoutTokens(data.location);
			}
			return data;
		};
	}
}

File: packages/clients/python/tangram/diagnostic.py
"""Diagnostic type and utilities.

This module mirrors the JavaScript packages/clients/js/src/diagnostic.ts.
"""

from __future__ import annotations

from typing import Literal, Optional, TypedDict

from tangram.location import Location, LocationData
from tangram.location import from_data as location_from_data
from tangram.location import to_data as location_to_data

Severity = Literal["error", "warning", "info", "hint"]


class Diagnostic(TypedDict, total=False):
    """A diagnostic message with optional location."""

    location: Optional[Location]
    message: str
    severity: Severity


class DiagnosticData(TypedDict, total=False):
    """Serialized diagnostic data."""

    location: Optional[LocationData]
    message: str
    severity: Severity


def to_data(value: Diagnostic) -> DiagnosticData:
    """Convert a diagnostic to data."""
    data: DiagnosticData = {
        "message": value["message"],
        "severity": value["severity"],
    }
    if value.get("location") is not None:
        data["location"] = location_to_data(value["location"])
    return data


def from_data(data: DiagnosticData) -> Diagnostic:
    """Convert data to a diagnostic."""
    diagnostic: Diagnostic = {
        "message": data["message"],
        "severity": data["severity"],
    }
    if data.get("location") is not None:
        diagnostic["location"] = location_from_data(data["location"])
    return diagnostic
